### ISSUE TITLE
Metal SSD streaming: layer-batched decode can read reused expert-cache buffers (deterministic wrong logits under cache pressure)

### ISSUE THREAD

**Issue by rinaldofesta** [antirez/ds4#384]
Following up on the `q2 logprob test fails on Metal+ssd-streaming` report in #16 — root cause isolated. Apple M5 Max 128GB, macOS Darwin 25.5, main `91bafb5`, `DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix.gguf`.

## Repro

```
DS4_TEST_SSD_STREAMING=1 DS4_TEST_SSD_STREAMING_CACHE_GB=16 ./ds4_test --logprob-vectors
```

`short_code_completion` step 1 flips its argmax deterministically (official `c`; local ranks `C` at -0.63 over `c` at -0.76). At `CACHE_GB=20` the same step is confidently wrong (`C` -0.11 vs `c` -2.24, ~2.1 nats). Bit-identical across repeated runs.

## Experimental matrix

| config (16GB cache unless noted) | result |
|---|---|
| full residency / auto budget (72.5GiB) / cold-start auto | OK |
| cache 24 / 28 / 32 / 48 GB | OK |
| cache 16 / 20 GB | **ERR, deterministic** |
| 16GB + `DS4_METAL_MOE_ONE_STAGE_PROFILE=1` | OK (but confounded: also flushes per stage) |
| 16GB + `DS4_METAL_DISABLE_STREAMING_EXPERT_MASKED_ADDR=1` | ERR, bit-identical |
| 16GB + `DS4_METAL_DISABLE_STREAMING_EXPERT_EARLY_LOAD=1` | ERR, bit-identical |
| 16GB + extra `wait_pending_command_buffers` before the missing-expert load | ERR, bit-identical |
| 16GB + `DS4_METAL_DISABLE_STREAMING_STATIC_MAP_STATE_CACHE=1` | ERR, bit-identical |
| 16GB + `DS4_METAL_DISABLE_STREAMING_STATIC_DECODE_MAP=1` | **OK** |
| 16 and 20GB + `DS4_METAL_DISABLE_STREAMING_LAYER_BATCH=1` | **OK** |

Also: prefill is clean — `--local-golden-vectors` reports identical metrics (down to `top20_max_abs=3.2229`) for streaming-16GB vs full residency. The divergence appears only at decode steps. CPU-side consistency checks at encode time (addr-table entries vs cache entries, buffer bytes vs model file bytes for all selected experts) all pass — the corruption is not observable at any single CPU checkpoint.

## Mechanism

With the static decode map, streaming decode uses the layer-batched path (`metal_graph_stream_decode_layer_batch_enabled`, ds4.c:18378+): all layers of a token are **encoded into one batched command buffer before the GPU executes any of them**. Expert-cache misses are serviced CPU-side during encoding, and under cache pressure `ds4_gpu_stream_expert_cache_take_reusable` (ds4_metal.m:9396) evicts a victim and reuses its slab buffers immediately. The protection passed down (`protect_layer` + current `selected_ids`) only covers the **layer currently being encoded** — an expert loaded for an *earlier layer of the same not-yet-executed batch* is a legal victim.

So under pressure: layer `i`'s dispatch is encoded referencing buffer B for expert E; while encoding layer `j > i`, E is evicted and B is refilled with expert E′'s bytes; the GPU then executes layer `i` reading E′'s weights as E. `useResource` keeps B resident, so nothing faults — the kernel just reads the wrong expert. This explains every row of the matrix:

- deterministic: encode order and eviction order are deterministic for a fixed prompt;
- cache-pressure-keyed: with ≥24GB (~3640 slots here) a token's working set survives its own encode window, so no intra-batch reuse;
- decode-only: prefill doesn't use the decode layer batch;
- insensitive to split/masked-addr/early-load toggles: the reuse happens regardless of which expert dispatch flavor is encoded;
- "fixed" by `MOE_ONE_STAGE_PROFILE`: its per-stage `end_commands`/`begin_commands` force execution before later loads can reuse anything;
- invisible to CPU-side checks: bookkeeping is self-consistent at every instant; the inconsistency is only between GPU-execution time and a later CPU write.

The drift magnitude depends on which experts get clobbered, which is why 20GB looks *worse* than 16GB despite less pressure. Plausibly the same mechanism affects the ROCm leg reported in #16 if the shared layer-batch path drives it — I can't test that here (no hardware).

## Workaround and cost

`DS4_METAL_DISABLE_STREAMING_LAYER_BATCH=1` restores correctness at every cache size tested. Measured cost at 16GB cache (short prompt, greedy, `-n 64`): generation 13.56 → 12.37 t/s (~-9%), prefill unchanged — so the batching is worth keeping and the fix should preserve it.

## Fix direction

Smallest shape I see, single fixed behavior, no knobs: extend eviction protection to every cache entry referenced by the current **unexecuted batch** — e.g. mark entries with a batch sequence number when their buffers are encoded (`useResource`'d), treat marked entries as non-reusable in `take_reusable`/`prepare_load_buffers`, and clear the marks when the batch is flushed-and-waited. If no victim exists outside the in-flight set, flush + wait once and retry — that degrades to today's eviction behavior only under extreme pressure instead of corrupting. Happy to build that PR with the matrix above as the red→green proof, if useful.

This is also a concrete instance of the class discussed above re: the decode-drift gate — every in-tree test except a near-tie argmax assert stays green while decode logits drift.

**Comment by antirez** [comment]
Thank you, fixed in 8384adf0

**Comment by rinaldofesta** [comment]
Verified on the reporting machine (M5 Max 128GB) at `8384adf`:

- `./ds4_test --metal-ssd-streaming-cache-pressure` → OK
- full `--logprob-vectors` with 16GB cache → OK (was ERR)
- full `--logprob-vectors` with 20GB cache → OK (was ERR, confident wrong token)
- layer batching confirmed still active and worth it: gen 19.35 t/s vs 17.29 t/s with `DS4_METAL_DISABLE_STREAMING_LAYER_BATCH=1` (16GB cache, short prompt, greedy)

Thanks for the fast fix — and for locking the repro in as `--metal-ssd-streaming-cache-pressure`.
